### ISSUE TITLE
prefix all modules by plc

### ISSUE THREAD

**Issue by catwell** [philanc/plc#7]
Currently, plc does not use a namespace. This is a proposal to move all Lua files to the "plc" namespace, i.e. create a directory called "plc" and put them inside it, then write for instance `require "plc.md5"` instead of `require "md5"`.

The goal of this is to solve conflicts with other packages which have the same name as one of the modules, md5 being a good example. It is also a prerequisite to package plc as a Lua rock.

**Comment by catwell** [comment]
If you want an idea of how it looks, I have [done that on my master branch](https://github.com/catwell/plc/commit/239fbc5e151ee233ee6e32c349d3100828a60974), as well as [fixed everything reported by luacheck](https://github.com/catwell/plc/commit/71711d047f55afcc9bb9bcc4bab65b0657f33629). Not sure if you want to do such changes on your repo though. I am thinking it would be useful to have a plc rock, add Travis integration for CI, etc. I can do this but please tell me if you think it's a bad idea.

**Comment by philanc** [comment]
Re luacheck:  Wow, very interesting. I have to learn about luacheck. Some things are obviously good (spaces, end of lines)..Some may require a bit of review. Thanks for the heads-up!

Re luarock:  Good idea, since people expect it. on my side, I don't use it. I tried to provide rockspecs in other repositories, but I am never sure my rockspec points to the master, or some tagged version and what it should do :-) and how to deal with luarocks.org   -- you certainly know better!

Re Travis:  I never used it. Maybe kind of overkill for a small project with sparse updates - but maybe an opportunity to learn how it works.  What would be the effort?

Re plc namespace - I have mixed feelings on this one. On one hand, I don't want plc be a structured library where you basically have to (or feel you have to) install all of it. I really like the idea of code snippets, no string attached.  

On the other hand, the more we go into *constructs* (eg. box.lua,  aead_chacha_poly.lua), the less the "snippet approach" works.  And a library with luarock support might be what many people are more comfortable with...  Hmmm.

Let me think a bit over it :-)

**Comment by catwell** [comment]
> Re luacheck: Wow, very interesting. I have to learn about luacheck. Some things are obviously good (spaces, end of lines)..Some may require a bit of review. Thanks for the heads-up!

Yes, the most important thing is probably the use of globals instead of local variables in some places.

>  Re luarock: Good idea, since people expect it. on my side, I don't use it. I tried to provide rockspecs in other repositories, but I am never sure my rockspec points to the master, or some tagged version and what it should do :-) and how to deal with luarocks.org -- you certainly know better!

I can help you with that yes. We would need the namespace thing first though. Otherwise we can go for a rock for every single file, but it's a bit overkill.

> Re Travis: I never used it. Maybe kind of overkill for a small project with sparse updates - but maybe an opportunity to learn how it works. What would be the effort?

It's not hard to set up so I think it could be nice. Not necessarily a must-have.

Re. the namespace, I was thinking about the "constructs" thing as well, plus I've had the tests fail because it required another MD5 module (I fixed that with a `package.path` change, but it made me think). It certainly does not prevent the user from using a single file or just a few of them, but you are right that it could make it look like it's not the case. Your call.

**Comment by philanc** [comment]
Ok, so I set up my mind about the namespace. For luacheck, I merged all the modifications (and the rc file) as-is and will take some time to play with it on the side. 

Thanks for all the work!

**Comment by catwell** [comment]
OK, this is awesome. I can add a SCM rockspec and Travis support now if you want.

(A SCM rockspec is a rockspec that pulls from the GitHub master branch directly, it is useful to test development releases, as opposed to normal "versioned" rockspecs that correspond to releases.)

**Comment by philanc** [comment]
For the SCM rockspec, it would be great. 

For versioned rockspecs, I wonder if they are even needed, given the nature of the project. Maybe before large restructuring, such as the plc name space?  what do you think?

Re Travis, I really have no idea. Would I need to _do_something?  :-)

**Comment by catwell** [comment]
It does not make much sense to make a rockspec before the PLC name change.

For Travis, the only thing you need to do is go to https://travis-ci.org, sign in with your GitHub account and activate Travis for this repository (by "flipping the switch", it's should be easy to find).

**Comment by philanc** [comment]
> It does not make much sense to make a rockspec before the PLC name change.

Of course. What I meant is: *in the future*, a rockspec for a tagged version might make sense before a large modification --such as for example the plc namespace restructuring.  Even in that case, I am not sure it would be so useful...

A side question:  If *I* open an issue on this philanc/plc repo, do you get a notification?  should I open an issue on your catwell/plc?

**Comment by catwell** [comment]
> If I open an issue on this philanc/plc repo, do you get a notification? should I open an issue on your catwell/plc?

Now yes, I will (because I started watching the repository). So no need to open it on my fork.

**Comment by philanc** [comment]
Good.
